File: examples/ruby/crud_integration_spec.rb
# frozen_string_literal: true

# RSpec integration test for the Pescheck Ruby client.
#
# Runs the full v2 CRUD lifecycle against a real Pescheck API using an OAuth2
# bearer token supplied via PESCHECK_ACCESS_TOKEN. There are NO mocks: every
# expectation is checked against a live response.
#
# Environment:
#   PESCHECK_BASE_URL      API base URL (default https://api-staging.pescheck.io)
#   PESCHECK_ACCESS_TOKEN  OAuth2 client-credentials access token (required)
#   PESCHECK_TEST_EMAIL    candidate email (default noreply@pescheck.nl)
#
# Without PESCHECK_ACCESS_TOKEN the whole suite is skipped (green, exit 0).
#
# Run:
#   cd examples/ruby
#   rspec -I../../clients/ruby/lib crud_integration_spec.rb

require 'uri'
require 'securerandom'
require 'pescheck-client'

# The lifecycle steps build on each other (a screening needs the profile created
# a step earlier), so they must run in source order rather than RSpec's default
# random order.
RSpec.describe 'Pescheck client CRUD lifecycle (live API)', order: :defined do
  # Profile/webhook names must be unique per run: the API soft-deletes and
  # returns a 500 on a duplicate name, so we suffix names and the webhook URL.
  # The division has no delete endpoint, so we reuse a fixed-named one.
  SUFFIX = SecureRandom.hex(4)
  DIVISION_NAME = 'E2E CI division'

  before(:all) do
    skip 'PESCHECK_ACCESS_TOKEN not set' if ENV['PESCHECK_ACCESS_TOKEN'].to_s.empty?

    base_url = ENV.fetch('PESCHECK_BASE_URL', 'https://api-staging.pescheck.io')
    @test_email = ENV.fetch('PESCHECK_TEST_EMAIL', 'noreply@pescheck.nl')
    uri = URI.parse(base_url)

    Pescheck.configure do |config|
      config.scheme = uri.scheme
      config.host = uri.host + (uri.port && uri.default_port != uri.port ? ":#{uri.port}" : '')
      config.base_path = uri.path
      config.access_token = ENV['PESCHECK_ACCESS_TOKEN']
    end

    @checks = Pescheck::ChecksApi.new
    @profiles = Pescheck::ProfilesApi.new
    @screenings = Pescheck::ScreeningsApi.new
    @webhooks = Pescheck::WebhooksApi.new
    @divisions = Pescheck::DivisionsApi.new

    # Shared mutable state. Instance variables set inside an example are not
    # visible to later examples or to after(:all), but mutating a hash created
    # here (and shared by reference with every example) is.
    @state = {}
  end

  after(:all) do
    # best-effort cleanup of the profile we created
    if @state && @state[:profile_id]
      begin
        @profiles.v2_profiles_destroy(@state[:profile_id])
      rescue StandardError
        # ignore cleanup failures
      end
    end
  end

  it 'lists check types and retrieves one' do
    check_list = @checks.v2_checks_list
    expect(check_list).not_to be_empty

    check_type = check_list[0].check_type
    @state[:check_type] = check_type

    retrieved = @checks.v2_checks_retrieve(check_type)
    expect(retrieved.check_type).to eq(check_type)
  end

  it 'creates, retrieves, patches and lists a profile' do
    created = @profiles.v2_profiles_create(
      Pescheck::V2ProfileCreate.new(
        name: "E2E test profile #{SUFFIX}",
        description: 'created by e2e',
        checks: [Pescheck::V2ProfileCheck.new(check_type: @state[:check_type])]
      )
    )
    @state[:profile_id] = created.id
    expect(@state[:profile_id]).not_to be_nil

    retrieved = @profiles.v2_profiles_retrieve(@state[:profile_id])
    expect(retrieved.id).to eq(@state[:profile_id])
    expect(retrieved.name).to eq("E2E test profile #{SUFFIX}")

    patched = @profiles.v2_profiles_partial_update(
      @state[:profile_id],
      patched_v2_profile_partial_update: Pescheck::PatchedV2ProfilePartialUpdate.new(description: 'updated by e2e')
    )
    expect(patched.description).to eq('updated by e2e')

    listed = @profiles.v2_profiles_list
    ids = (listed.respond_to?(:results) ? listed.results : listed).map(&:id)
    expect(ids).to include(@state[:profile_id])
  end

  it 'fully updates a profile via PUT' do
    # PUT replaces the whole profile: existing checks must be referenced by their
    # profile_check_id (the check entry's id), otherwise the API rejects a second
    # check of the same type.
    current = @profiles.v2_profiles_retrieve(@state[:profile_id])
    checks = current.checks.map do |entry|
      Pescheck::V2ProfileUpdateCheck.new(check_type: entry.check_type, profile_check_id: entry.id)
    end

    new_name = "E2E renamed profile #{SUFFIX}"
    updated = @profiles.v2_profiles_update(
      @state[:profile_id],
      Pescheck::V2ProfileUpdate.new(
        name: new_name,
        description: 'put-updated by e2e',
        checks: checks
      )
    )
    expect(updated.name).to eq(new_name)
    expect(updated.description).to eq('put-updated by e2e')

    retrieved = @profiles.v2_profiles_retrieve(@state[:profile_id])
    expect(retrieved.name).to eq(new_name)
  end

  it 'creates a screening, retrieves it and lists its documents' do
    screening = @screenings.v2_screenings_create(
      Pescheck::V2ScreeningCreate.new(
        profile_id: @state[:profile_id],
        candidate: Pescheck::V2Candidate.new(
          first_name: 'E2E', last_name: 'Tester', email: @test_email
        )
      )
    )
    expect(screening.id).not_to be_nil
    expect(screening.status).not_to be_nil
    @state[:screening_id] = screening.id

    retrieved = @screenings.v2_screenings_retrieve(screening.id)
    expect(retrieved.id).to eq(screening.id)
    expect(retrieved.candidate.email).to eq(@test_email)
    expect(retrieved.candidate.first_name).to eq('E2E')

    # documents list should return without error (may be empty)
    expect { @screenings.v2_screenings_documents_list(screening.id) }.not_to raise_error
  end

  it 'lists screenings and finds the one just created' do
    listed = @screenings.v2_screenings_list
    items = listed.respond_to?(:results) ? listed.results : listed
    expect(items).not_to be_empty
    expect(items.map(&:id)).to include(@state[:screening_id])
  end

  it 'creates, lists and deletes a webhook' do
    hook = @webhooks.create_webhook2(
      Pescheck::Webhook.new(
        name: "E2E webhook #{SUFFIX}",
        url: "https://example.com/e2e-hook-#{SUFFIX}",
        events: ['screening.status_changed']
      )
    )
    expect(hook.id).not_to be_nil
    expect(hook.url).to eq("https://example.com/e2e-hook-#{SUFFIX}")
    expect(hook.events).to include('screening.status_changed')

    listed = @webhooks.list_webhooks2
    expect(listed).not_to be_nil

    # webhook id deserializes as a UUID; the delete path param wants a string.
    expect { @webhooks.delete_webhook2(hook.id.to_s) }.not_to raise_error
  end

  it 'lists divisions, reuses-or-creates one and patches it' do
    existing = @divisions.v2_organisations_divisions_list.results
    div = existing.find { |d| d.name == DIVISION_NAME }

    if div.nil?
      div = @divisions.v2_organisations_divisions_create(
        Pescheck::DivisionWrite.new(
          name: DIVISION_NAME, city: 'Amsterdam', address: 'Teststraat 1',
          postal: '1011AA', phone: '+31200000000', contact_name: 'E2E',
          contact_email: 'e2e@example.com', invoice_email: 'e2e@example.com'
        )
      )
    end
    expect(div.id).not_to be_nil
    expect(div.name).to eq(DIVISION_NAME)
    @state[:division_id] = div.id

    patched = @divisions.v2_organisations_divisions_partial_update(
      div.id,
      patched_division_write: Pescheck::PatchedDivisionWrite.new(city: 'Rotterdam')
    )
    expect(patched.city).to eq('Rotterdam')
  end

  it 'retrieves a division by id' do
    retrieved = @divisions.v2_organisations_divisions_retrieve(@state[:division_id])
    expect(retrieved.id).to eq(@state[:division_id])
    expect(retrieved.name).to eq(DIVISION_NAME)
    expect(retrieved.city).to eq('Rotterdam')
  end

  it 'fully updates a division via PUT' do
    updated = @divisions.v2_organisations_divisions_update(
      @state[:division_id],
      Pescheck::DivisionWrite.new(
        name: DIVISION_NAME, city: 'Utrecht', address: 'Teststraat 2',
        postal: '3511AA', phone: '+31300000000', contact_name: 'E2E PUT',
        contact_email: 'e2e-put@example.com', invoice_email: 'e2e-put@example.com'
      )
    )
    expect(updated.city).to eq('Utrecht')
    expect(updated.address).to eq('Teststraat 2')
    expect(updated.contact_name).to eq('E2E PUT')

    retrieved = @divisions.v2_organisations_divisions_retrieve(@state[:division_id])
    expect(retrieved.city).to eq('Utrecht')
  end
end

# Negative cases need a configured client but not the CRUD lifecycle state, so
# they live in their own describe block.
RSpec.describe 'Pescheck client error handling (live API)', order: :defined do
  before(:all) do
    skip 'PESCHECK_ACCESS_TOKEN not set' if ENV['PESCHECK_ACCESS_TOKEN'].to_s.empty?

    @base_url = ENV.fetch('PESCHECK_BASE_URL', 'https://api-staging.pescheck.io')
    uri = URI.parse(@base_url)

    Pescheck.configure do |config|
      config.scheme = uri.scheme
      config.host = uri.host + (uri.port && uri.default_port != uri.port ? ":#{uri.port}" : '')
      config.base_path = uri.path
      config.access_token = ENV['PESCHECK_ACCESS_TOKEN']
    end
  end

  it 'raises a 401 ApiError when the access token is invalid' do
    uri = URI.parse(@base_url)
    bogus = Pescheck::Configuration.new.tap do |config|
      config.scheme = uri.scheme
      config.host = uri.host + (uri.port && uri.default_port != uri.port ? ":#{uri.port}" : '')
      config.base_path = uri.path
      config.access_token = 'bogus-invalid-token'
    end
    checks = Pescheck::ChecksApi.new(Pescheck::ApiClient.new(bogus))

    expect { checks.v2_checks_list }.to raise_error(Pescheck::ApiError) do |error|
      expect(error.code).to eq(401)
    end
  end

  it 'raises a 404 ApiError when retrieving a non-existent profile' do
    profiles = Pescheck::ProfilesApi.new
    missing_id = SecureRandom.uuid

    expect { profiles.v2_profiles_retrieve(missing_id) }.to raise_error(Pescheck::ApiError) do |error|
      expect(error.code).to eq(404)
    end
  end
end

File: examples/javascript/test/crud.integration.spec.js
// Integration test for the Pescheck JavaScript client.
//
// Exercises the full v2 CRUD surface against the real Pescheck API. No mocks:
// it authenticates with an OAuth2 bearer token and self-cleans where possible.
//
// The suite is skipped (reported pending, exit 0) unless PESCHECK_ACCESS_TOKEN
// is set, so it stays green without credentials.
//
// Environment variables:
//   PESCHECK_BASE_URL      - API base URL (default https://api-staging.pescheck.io)
//   PESCHECK_ACCESS_TOKEN  - OAuth2 client-credentials access token (required to run)
//   PESCHECK_TEST_EMAIL    - candidate email (default noreply@pescheck.nl)

const assert = require('node:assert');
const crypto = require('node:crypto');
const PescheckApi = require('@pescheckit/pescheck-client-js');

// Profile/webhook names must be unique per run: the API soft-deletes and returns
// a 500 on a duplicate name, so we suffix them (and the webhook URL). The division
// has no delete endpoint, so we reuse a fixed-named one instead.
const SUFFIX = crypto.randomBytes(4).toString('hex');
const DIVISION_NAME = 'E2E CI division';

describe('Pescheck client CRUD lifecycle', function () {
  // Real network calls: give each step generous headroom.
  this.timeout(60000);

  const baseUrl = process.env.PESCHECK_BASE_URL || 'https://api-staging.pescheck.io';
  const testEmail = process.env.PESCHECK_TEST_EMAIL || 'noreply@pescheck.nl';

  let checks, profiles, screenings, webhooks, divisions;
  let checkType;
  let profileId = null;
  let screeningId = null;
  let divId = null;

  before(function () {
    if (!process.env.PESCHECK_ACCESS_TOKEN) {
      this.skip();
    }

    // Configure the shared default client: base path + OAuth2 bearer access token.
    const defaultClient = PescheckApi.ApiClient.instance;
    defaultClient.basePath = baseUrl;
    defaultClient.authentications['oauth2'].accessToken = process.env.PESCHECK_ACCESS_TOKEN;

    checks = new PescheckApi.ChecksApi();
    profiles = new PescheckApi.ProfilesApi();
    screenings = new PescheckApi.ScreeningsApi();
    webhooks = new PescheckApi.WebhooksApi();
    divisions = new PescheckApi.DivisionsApi();
  });

  after(async function () {
    // Best-effort cleanup of the profile we created.
    if (profileId) {
      try {
        await profiles.v2ProfilesDestroy(profileId);
      } catch (e) {
        // ignore
      }
    }
  });

  it('lists and retrieves checks', async function () {
    const checkList = await checks.v2ChecksList();
    assert.ok(checkList.length > 0, 'expected at least one check type');
    checkType = checkList[0].check_type;
    const retrieved = await checks.v2ChecksRetrieve(checkType);
    assert.strictEqual(retrieved.check_type, checkType);
  });

  it('creates, retrieves and patches a profile', async function () {
    const name = `E2E test profile ${SUFFIX}`;
    const created = await profiles.v2ProfilesCreate({
      name,
      description: 'created by e2e',
      checks: [{ check_type: checkType }],
    });
    profileId = created.id;
    assert.ok(profileId, 'expected a profile id');

    const retrieved = await profiles.v2ProfilesRetrieve(profileId);
    assert.strictEqual(retrieved.id, profileId);
    assert.strictEqual(retrieved.name, name, 'profile should keep the name it was created with');

    const patched = await profiles.v2ProfilesPartialUpdate(profileId, {
      patchedV2ProfilePartialUpdate: { description: 'updated by e2e' },
    });
    assert.strictEqual(patched.description, 'updated by e2e');
    assert.strictEqual(patched.name, name, 'PATCH of description should leave the name unchanged');

    const list = await profiles.v2ProfilesList();
    const items = Array.isArray(list) ? list : list.results;
    assert.ok(items.some((p) => p.id === profileId), 'profile should appear in list');
  });

  it('replaces a profile with PUT', async function () {
    // PUT replaces the whole profile (name + description + checks). Existing
    // checks must be referenced by their profile_check_id to be updated in
    // place; sending a bare check_type the profile already has is rejected. A
    // duplicate name would 500, so the replacement name keeps the unique suffix.
    const current = await profiles.v2ProfilesRetrieve(profileId);
    const checksUpdate = current.checks.map((c) => ({
      check_type: c.check_type,
      profile_check_id: c.id,
    }));
    const newName = `E2E test profile (PUT) ${SUFFIX}`;
    const updated = await profiles.v2ProfilesUpdate(profileId, {
      name: newName,
      description: 'replaced by e2e PUT',
      checks: checksUpdate,
    });
    assert.strictEqual(updated.id, profileId);
    assert.strictEqual(updated.name, newName, 'PUT should replace the profile name');
    assert.strictEqual(updated.description, 'replaced by e2e PUT');
    assert.ok(updated.checks.some((c) => c.check_type === checkType), 'PUT should keep the check');
  });

  it('creates and retrieves a screening with documents', async function () {
    const screening = await screenings.v2ScreeningsCreate({
      profile_id: profileId,
      candidate: { first_name: 'E2E', last_name: 'Tester', email: testEmail },
    });
    assert.ok(screening.status, 'expected a screening status');
    screeningId = screening.id;

    const retrieved = await screenings.v2ScreeningsRetrieve(screening.id);
    assert.strictEqual(retrieved.id, screening.id);
    assert.ok(retrieved.candidate, 'expected the screening to carry a candidate');
    assert.strictEqual(retrieved.candidate.email, testEmail, 'candidate email should round-trip');

    await screenings.v2ScreeningsDocumentsList(screening.id);
  });

  it('lists screenings and finds the created screening', async function () {
    const list = await screenings.v2ScreeningsList();
    const items = Array.isArray(list) ? list : list.results;
    assert.ok(items.length > 0, 'expected at least one screening in the list');
    assert.ok(
      items.some((s) => s.id === screeningId),
      'the screening we created should appear in the list'
    );
  });

  it('creates, lists and deletes a webhook', async function () {
    const url = `https://example.com/e2e-hook-${SUFFIX}`;
    const events = ['screening.status_changed'];
    const hook = await webhooks.createWebhook2({
      name: `E2E webhook ${SUFFIX}`,
      url,
      events,
    });
    assert.ok(hook.id, 'expected a webhook id');
    assert.strictEqual(hook.url, url, 'webhook url should round-trip');
    assert.deepStrictEqual(hook.events, events, 'webhook events should round-trip');

    const list = await webhooks.listWebhooks2();
    const items = Array.isArray(list) ? list : list.results;
    assert.ok(items.some((w) => String(w.id) === String(hook.id)), 'webhook should appear in list');

    await webhooks.deleteWebhook2(String(hook.id));
  });

  it('lists, reuses-or-creates and patches a division', async function () {
    const existing = (await divisions.v2OrganisationsDivisionsList()).results;
    let div = existing.find((d) => d.name === DIVISION_NAME);
    if (!div) {
      div = await divisions.v2OrganisationsDivisionsCreate({
        name: DIVISION_NAME, city: 'Amsterdam', address: 'Teststraat 1',
        postal: '1011AA', phone: '+31200000000', contact_name: 'E2E',
        contact_email: 'e2e@example.com', invoice_email: 'e2e@example.com',
      });
    }
    assert.ok(div.id, 'expected a division id');
    assert.strictEqual(div.name, DIVISION_NAME, 'division should carry the expected name');
    divId = div.id;

    const patched = await divisions.v2OrganisationsDivisionsPartialUpdate(div.id, {
      patchedDivisionWrite: { city: 'Rotterdam' },
    });
    assert.strictEqual(patched.city, 'Rotterdam');
  });

  it('retrieves and replaces a division with PUT', async function () {
    const retrieved = await divisions.v2OrganisationsDivisionsRetrieve(divId);
    assert.strictEqual(retrieved.id, divId);
    assert.strictEqual(retrieved.name, DIVISION_NAME, 'retrieved division should keep its name');
    // The previous PATCH set the city to Rotterdam.
    assert.strictEqual(retrieved.city, 'Rotterdam', 'retrieve should reflect the prior PATCH');

    // PUT requires the full DivisionWrite body; reuse the retrieved fields and
    // change the city so we can assert the replacement took effect.
    const updated = await divisions.v2OrganisationsDivisionsUpdate(divId, {
      name: DIVISION_NAME, city: 'Utrecht', address: retrieved.address,
      postal: retrieved.postal, phone: retrieved.phone,
      contact_name: retrieved.contact_name, contact_email: retrieved.contact_email,
      invoice_email: retrieved.invoice_email,
    });
    assert.strictEqual(updated.city, 'Utrecht', 'PUT should replace the division city');
    assert.strictEqual(updated.name, DIVISION_NAME);
  });

  it('rejects unauthorized requests with 401', async function () {
    // A fresh client carrying a bogus OAuth2 token must be refused. superagent
    // surfaces the HTTP status on the rejected error as err.status.
    const badClient = new PescheckApi.ApiClient();
    badClient.basePath = baseUrl;
    badClient.authentications['oauth2'].accessToken = 'definitely-not-a-valid-token';
    const unauthChecks = new PescheckApi.ChecksApi(badClient);

    await assert.rejects(
      () => unauthChecks.v2ChecksList(),
      (err) => {
        assert.strictEqual(err.status, 401, `expected 401, got ${err.status}`);
        return true;
      }
    );
  });

  it('rejects retrieving an unknown profile with 404', async function () {
    const missingId = crypto.randomUUID();
    await assert.rejects(
      () => profiles.v2ProfilesRetrieve(missingId),
      (err) => {
        assert.strictEqual(err.status, 404, `expected 404, got ${err.status}`);
        return true;
      }
    );
  });
});
